spot-vaults/contracts/UsdcSpotManager.sol

```diff
@@ -0,0 +1,228 @@
+        return r;
+    }
+
+    //--------------------------------------------------------------------------
+    // External write methods
+
+    /// @notice Executes vault rebalance.
+    function rebalance() public {
+        (uint256 deviation, bool deviaitonValid) = computeDeviationFactor();
+
+        // Update limit range.
+        bool inNarrowLimitRange = checkNarrowLimitRange(deviation);
+        bool prevInNarrowLimitRange = (VAULT.limitThreshold() <= limitThresholdNarrow);
+        bool limitRangeUpdate = (prevInNarrowLimitRange != inNarrowLimitRange);
+        if (limitRangeUpdate) {
+            (deviaitonValid && inNarrowLimitRange)
+                ? VAULT.setLimitThreshold(limitThresholdNarrow)
+                : VAULT.setLimitThreshold(limitThresholdWide);
+        }
+
+        // Execute rebalance.
+        // NOTE: the vault.rebalance() will revert if enough time has not elapsed.
+        // We thus override with a force rebalance.
+        // https://learn.charm.fi/charm/technical-references/core/alphaprovault#rebalance
+        limitRangeUpdate ? _execForceRebalance() : VAULT.rebalance();
+    }
+
+    /// @notice Computes the deviation between SPOT's market price and it's FMV price.
+    /// @return The computed deviation factor.
+    function computeDeviationFactor() public returns (uint256, bool) {
+        uint256 spotMarketPrice = getSpotUSDPrice();
+        (uint256 spotTargetPrice, bool spotTargetPriceValid) = spotAppraiser.perpPrice();
+        (, bool usdcPriceValid) = spotAppraiser.usdPrice();
+        bool deviaitonValid = (spotTargetPriceValid && usdcPriceValid);
+        uint256 deviation = spotTargetPrice > 0
+            ? FullMath.mulDiv(spotMarketPrice, ONE, spotTargetPrice)
+            : type(uint256).max;
+        deviation = (deviation > MAX_DEVIATION) ? MAX_DEVIATION : deviation;
+        return (deviation, deviaitonValid);
+    }
+
+    //-----------------------------------------------------------------------------
+    // External Public view methods
+
+    /// @notice Checks if the vault is in the narrow limit range based on the current deviation factor.
+    function checkNarrowLimitRange(uint256 deviation) public view returns (bool) {
+        int24 _marketPrice = VAULT.getTwap();
+        int24 _limitLower = VAULT.limitLower();
+        int24 _limitUpper = VAULT.limitUpper();
+        int24 _limitPrice = (_limitLower + _limitUpper) / 2;
+        // The limit range has more SPOT than USDC if `_marketPrice >= _limitPrice`,
+        // so the vault looks to sell SPOT.
+        // NOTE: This assumes that in the underlying univ3 pool
+        // token0 is USDC and token1 is SPOT.
+        bool isSpotAsk = (_marketPrice >= _limitPrice);
+        return ((deviation >= ONE && isSpotAsk) || (deviation <= ONE && !isSpotAsk));
+    }
+
+    /// @return The computed SPOT price in USD from the underlying univ3 pool.
+    function getSpotUSDPrice() public view returns (uint256) {
```

Comment:
This is perhaps the first time any code we've written relies on realtime market prices from dexes.

I think the security risk is minimal, since it would require an attacker to manipulate the spot market to profit from forcing limit range changes, right? Not sure I see any benefits, but calling out just in case.

Reply:
We will be using a 1hr TWAP here. The wampl vault has been configured to use a 1hr twap, the spot vault needs to be updated.